docs/source/architecture.rst

```diff
@@ -1,69 +1,80 @@
 Architecture
 ============
 
+It is recommended that the following is read in conjunction with exploring
+the codebase. `dialect_ansi.py` in particular is helpful to understand the
+recursive structure of segments and grammars.
+
 Stage 1, the templater
 ----------------------
 
-The templater takes raw files and optionally fills in any configurable
-sections before passing onto the linter. In particular most templating
-systems use curly brackets *{}* to indicate templatable sections and
-these are not currently set up in any of the lexers, and so will fail
-at the next step if not dealt with here.
+This stage only applies to templated SQL, most commonly Jinja and dbt. Vanilla
+SQL is sent straight to stage 2, the lexer.
+
+In order to lint templated SQL, SQLFluff must first convert the 'raw' or
+pre-templated code into valid SQL, which can then be parsed. The templater
+returns both the raw and post-templated SQL so that any rule violations which
+occur in templated sections can be ignored and the rest mapped to their
+original line location for user feedback.
 
 *SQLFluff* supports 2 templating engines:
 
 .. _Jinja: https://jinja.palletsprojects.com/
 .. _dbt: https://docs.getdbt.com/
 
-Jinja_ is the default templater used by *SQLFluff*. Under the covers, dbt_ also
-uses Jinja_ but in *SQLFluff* it is a separate templater which relies directly
-on the dbt_ compiler.
+Under the hood dbt_ also uses Jinja_, but in *SQLFluff* uses a separate
+mechanism which interfaces directly with the dbt_ python package.
 
 For more details on how to configure the templater see :ref:`templateconfig`.
 
 
 Stage 2, the lexer
 ------------------
 
-The lexer takes raw text and splits it into tokens. Nothing is *removed* but
-whitespace and code are separated. In principle all identifiers should be
-separated at this stage, but they should not be imparted any meaning at this
-stage. Any files which cannot be lexed, should raise a *SQLLexError*.
+The lexer takes valid SQL and separates it into segments of whitespace and
```

Comment:
Nitpick: Saying "valid SQL" seems potentially misleading, since sometimes SQL is invalid. Maybe say "SQL" or "presumably valid SQL"?

Reply:
Agreed, SQL it is!